### PR TITLE
Add self-service course participation exit (cancel/abort)

### PR DESCRIPTION
**Pull request by steffen74**
## Summary
Adds the ability for participants to self-service cancel or abort their course enrollment, with different options depending on whether the course has started.

## Key Changes
- **New `participationExit` module** (`participationExit.ts`): Core logic determining whether a participant can exit and which action to offer
  - `getParticipationExitKind()` returns `'CANCEL'` before course starts, `'ABORT'` while running, or `null` when unavailable
  - Handles edge cases: missing dates, unparsable timestamps, paid invoices, terminal statuses
  - Uses course session window (first start to last end) to determine timing

- **New `ParticipationExitButton` component**: UI for the exit action
  - Renders as a quiet text button under enrollment status (not competing with registration CTA)
  - Shows confirmation dialog before submitting status change
  - Handles loading/error states and calls `onExit` callback after success

- **New GraphQL mutation** `CANCEL_OWN_ENROLLMENT`: Updates enrollment status with guards
  - Only allows status change from eligible statuses (APPLIED, WAITLIST, INVITED, CONFIRMED, REGISTERED)
  - Prevents race conditions where enrollment moved to terminal state

- **Integration into `CourseContent` and `Registration` components**:
  - Wires up `ParticipationExitButton` in registration rail
  - Tracks participation exit state and refetches course data on exit
  - Passes `onParticipationExit` callback through component hierarchy

- **Comprehensive test coverage**:
  - `participationExit.test.ts`: 13 tests covering timing logic, edge cases, and status eligibility
  - `ParticipationExitButton.test.tsx`: 7 tests covering UI interactions, confirmation flow, and error handling

- **Localization**: Added German and English strings for cancel/abort actions and confirmation dialogs
  - Updated "cancelled" status label from "invitation rejected" to "registration cancelled" for clarity

## Notable Implementation Details
- Timezone-aware: Comparisons use absolute instants (`timestamptz`), not calendar days
- Graceful degradation: Missing/unparsable dates treated as "not yet started" to avoid locking participants out
- Refund-aware: Paid invoices prevent self-service exit (refunds out of scope)
- Race-condition safe: `affected_rows: 0` indicates enrollment moved on; caller refetches to show actual state

https://claude.ai/code/session_01HXTM6ESPPDLo3vvutsbn61

## Summary by CodeRabbit

* **New Features**
  * Participants can cancel enrollment before a course starts or abort participation while it is in progress.
  * Confirmation dialogs, loading states, success notifications, and error messages guide the process.
  * Exit options are unavailable after completion, for terminal statuses, or when an invoice has been paid.
  * Registration status updates reflect cancelled or aborted participation.

* **Documentation**
  * Added English and German translations for participation exit actions, confirmations, notifications, and errors.

### CODE REVIEW

**Comment by steffen74** [comment]
@coderabbit review

File: frontend-nx/apps/edu-hub/components/pages/CourseContent/Registration/participationExit.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  let start: number | null = null;
+  let end: number | null = null;
+
+  for (const session of sessions) {
+    const sessionStart = toTime(session.startDateTime);
+    const sessionEnd = toTime(session.endDateTime);
+
+    if (sessionStart != null && (start == null || sessionStart < start)) start = sessionStart;
+    for (const candidate of [sessionStart, sessionEnd]) {
+      if (candidate != null && (end == null || candidate > end)) end = candidate;
```
[the remark this answers is not in the document]
Comment (steffen74) [reply frontend-nx/apps/edu-hub/components/pages/CourseContent/Registration/participationExit.ts]: Confirmed and fixed in 895ac6a.

The fallback was meant to prevent exactly the lockout it caused: taking the latest start as the window end closed the window a millisecond after the last session began. `getCourseWindow` now returns `end: null` when there is no known end, or when a session starts after every end we do know, and `getParticipationExitKind` treats an open-ended window as still running.

Tests updated: the `10:01` case now expects `ABORT`, plus two new cases separating "only the last session lacks an end" (stays open) from "only an earlier one does" (still closes on the last real end time).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: frontend-nx/apps/edu-hub/components/pages/CourseContent/Registration/ParticipationExitButton.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+}) => {
+  const t = useTranslations('course');
+  const [isConfirming, setIsConfirming] = useState(false);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [hasFailed, setHasFailed] = useState(false);
+
+  const [cancelOwnEnrollment] = useRoleMutation<CancelOwnEnrollment, CancelOwnEnrollmentVariables>(
+    CANCEL_OWN_ENROLLMENT
+  );
+
```
[the remark this answers is not in the document]
Comment (steffen74) [reply frontend-nx/apps/edu-hub/components/pages/CourseContent/Registration/ParticipationExitButton.tsx:53]: Confirmed and fixed in 895ac6a. I verified the premise: `createStripeCheckout` creates a session without checking for an existing one, the webhook inserts an invoice per session, and there is no unique constraint on `courseEnrollmentId` — so `[newest: ISSUED, older: PAID]` is reachable after a retry on an already-paid enrollment.

Taken as suggested:

- `ENROLLMENT_FRAGMENT` drops `limit: 1` and keeps `order_by: { created_at: desc }`, so latest-only readers are unaffected. This does not change any generated type (GraphQL field arguments do not appear in the TS output), so no codegen churn.
- New `hasPaidInvoice(invoices)` in `utils/invoicePaymentStatus.ts` asks whether *any* invoice is `PAID`; `ParticipationExitButton` uses it instead of `getPaymentStatusFromInvoices(...) === 'COMPLETED'`.
- `getPaymentStatusFromInvoices` stays latest-only, as you noted `canRetryPayment` depends on it.
- Regression tests in `utils/invoicePaymentStatus.test.ts` cover `[ISSUED, PAID]` in both orders, and a component test asserts no exit button renders for that enrollment.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: frontend-nx/apps/edu-hub/locales/de.json [hunk 3]
```diff
@@ -630,13 +630,17 @@
       "register_now": "Jetzt registrieren",
       "accept_invitation": "Einladung annehmen",
       "placeholder_motivation_letter": "Meine Motivation ...",
-      "retry_payment": "Zahlung erneut versuchen"
+      "retry_payment": "Zahlung erneut versuchen",
+      "cancel_participation": "Anmeldung stornieren",
+      "abort_participation": "Teilnahme abbrechen",
+      "cancel_participation_confirm": "Möchtest du deine Anmeldung für „{title}“ wirklich stornieren? Dein Platz wird dann wieder frei.",
+      "abort_participation_confirm": "Möchtest du deine Teilnahme an „{title}“ wirklich abbrechen?"
```
[the remark this answers is not in the document]
Comment (steffen74) [reply frontend-nx/apps/edu-hub/locales/de.json]: Fixed in 895ac6a. I checked the cited rule before acting — `.claude/commands/frontend-patterns.md` does state `namespace (camelCase) > ComponentGroup (PascalCase) > key (snake_case)`, so the new keys were in the wrong shape even though the neighbouring `course.registration` / `course.modal` / `course.errors` groups predate it.

- `course.ParticipationExitButton.{cancel_participation, abort_participation, cancel_participation_confirm, abort_participation_confirm, exit_failed}`
- `course.CourseContent.{cancel_success_message, abort_success_message}`

Applied in both `de.json` and `en.json`, with callers and tests updated.

One key I deliberately left where it is: `course.status.cancelled`. This PR only rewords it (it read "you declined your invitation", from when declining an invitation was the only route to `CANCELLED`). Moving it would touch `RegistrationStatus.tsx`'s whole sibling set of status keys, which is beyond this change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: frontend-nx/apps/edu-hub/queries/insertEnrollment.ts [hunk 4]
```diff
@@ -229,3 +229,40 @@ export const UPDATE_ENROLLMENT_STATUS_FOR_INVITE = gql`
...
+ * `user_access` update permission filters on `userId = X-Hasura-User-Id`.
+ *
+ * Deliberately not UPDATE_ENROLLMENT_STATUS: that one also sets
+ * `invitationExpirationDate`, so calling it without `expire` NULLs a column
+ * this action has no business touching.
+ */
+export const CANCEL_OWN_ENROLLMENT = gql`
+  mutation CancelOwnEnrollment(
+    $enrollmentId: Int!
+    $status: CourseEnrollmentStatus_enum!
```
[the remark this answers is not in the document]
Comment (steffen74) [reply frontend-nx/apps/edu-hub/queries/insertEnrollment.ts:250]: The finding is correct, but it is not introduced here and I'd rather not fix it blind in this PR. Flagging for @steffen74 to decide.

**What I verified.** The `user_access` update permission on `CourseEnrollment` grants `status` with `filter: { userId: { _eq: X-Hasura-User-Id } }` and `check: {}`. That block is byte-identical on `develop`, and this PR touches no file under `backend/` — it consumes the existing grant, exactly as `Onboarding.tsx` (`INVITED → CONFIRMED` / `CANCELLED`) and the `UPDATE_ENROLLMENT` upsert already do. So the bypass is reachable on `develop` today, and it is wider than cancel/abort: a participant can already set their own enrollment to `COMPLETED`, which is the status certificates hang off.

The client-side window is documented as client-side in the PR's plan; it sits at the same level of enforcement as `isRegistrationClosed` and the application deadline, which are likewise unenforced server-side for logged-in users.

**What I'd propose, in two pieces:**

1. *Bounded, and the one I'd do first.* Add a `check` to the `user_access` update permission restricting the post-update row's `status` to what a participant may legitimately set — `APPLIED`, `WAITLIST`, `CONFIRMED`, `REGISTERED`, `CANCELLED`, `ABORTED`. That closes self-promotion to `COMPLETED`, `INVITED` and `REJECTED` without a new handler. It is expressible as a Hasura boolean expression.
2. *The heavy lift you describe.* Payment and course-window eligibility cannot go in a permission expression — Hasura cannot compare against an aggregate over `Session` — so they need a Hasura action or a Postgres trigger.

I've not pushed either, for two reasons: (1) is a shared permission that also gates the registration upsert and the onboarding confirm, so getting the allowed set wrong breaks enrolment itself; and I have no Hasura instance in this environment to apply the metadata against and verify those flows. A speculative metadata change is the wrong thing to put in a review round.

Suggest tracking both as a follow-up issue against the pre-existing permission rather than widening this PR. Happy to do (1) here instead if you'd prefer — say the word and I'll validate it against a running stack.

---
_Generated by [Claude Code](https://claude.ai/code)_
